README.md
# Realex Remote Java SDK
You can sign up for a Realex account at https://developer.realexpayments.com
## Requirements
Java 1.6 and later.
## Installation
### Maven users
Add this dependency to your project's POM:
```xml
<dependency>
  <groupId>com.realexpayments.remote.sdk</groupId>
  <artifactId>rxp-remote-java</artifactId>
  <version>1.3.1</version>
</dependency>
```

### Gradle users
Add this dependency to your project's build file:
```
compile "com.realexpayments.remote.sdk:rxp-remote-java:1.3.1"
```
## Usage

Please see https://developer.realexpayments.com for more comprehensive integration guides.

### Authorisation

```java
Card card = new Card()
	.addExpiryDate("0119")
	.addNumber("4242424242424242")
	.addType(CardType.VISA)
	.addCardHolderName("Joe Smith")
	.addCvn("123")
	.addCvnPresenceIndicator(PresenceIndicator.CVN_PRESENT);

PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
	.addType(PaymentType.AUTH)
	.addAmount(100)
	.addCurrency("EUR")
	.addCard(card)
	.addAutoSettle(new AutoSettle().addFlag(AutoSettleFlag.TRUE));

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Authorisation (With Address Verification)

```java
Card card = new Card()
	.addExpiryDate("0119")
	.addNumber("420000000000000000")
	.addType(CardType.VISA)
	.addCardHolderName("Joe Smith")
	.addCvn("123")
	.addCvnPresenceIndicator(PresenceIndicator.CVN_PRESENT);
 
PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
	.addType(PaymentType.AUTH)
	.addAmount(100)
	.addCurrency("EUR")
	.addCard(card)
	.addAutoSettle(new AutoSettle().addFlag(AutoSettleFlag.TRUE))
	.addAddressVerificationServiceDetails("382 The Road", "WB1 A42");

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Authorisation (Mobile)

```java
PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
	.addType(PaymentType.AUTH_MOBILE)
	.addAutoSettle(new AutoSettle().addFlag(AutoSettleFlag.TRUE))
	.addMobile("apple-pay")
	.addToken("{auth mobile payment token}");

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Settle

```java
PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
	.addType(PaymentType.SETTLE)
	.addOrderId("Order ID from original transaction")
	.addAmount(100)
	.addCurrency("EUR")
	.addPaymentsReference("pasref from original transaction");

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Void

```java
PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
	.addType(PaymentType.VOID)
	.addOrderId("Order ID from original transaction")
	.addPaymentsReference("pasref from original transaction");

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Rebate

```java
PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
	.addType(PaymentType.REBATE)
	.addOrderId("Order ID from original transaction")
	.addAmount(100)
	.addCurrency("EUR")
	.addPaymentsReference("pasref from original transaction")
	.addAuthCode("Auth code from original transaction")
	.addRefundHash("SHA1 hash of rebate password");

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### OTB

```java
Card card = new Card()
	.addExpiryDate("0119")
	.addNumber("420000000000000000")
	.addType(CardType.VISA)
	.addCardHolderName("Joe Smith")
	.addCvn("123")
	.addCvnPresenceIndicator(PresenceIndicator.CVN_PRESENT);
 
 PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
	.addType(PaymentType.OTB)
	.addCard(card);
	
RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);	
```

### Refund

```java
Card card = new Card()
	.addExpiryDate("0119")
	.addNumber("420000000000000000")
	.addType(CardType.VISA)
	.addCardHolderName("Joe Smith")
	.addCvn("123")
	.addCvnPresenceIndicator(PresenceIndicator.CVN_PRESENT);
	
PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
	.addType(PaymentType.REFUND)
	.addAmount(100)
	.addCurrency("EUR")
	.addCard(card)
	.addRefundHash("SHA1 hash of refund password");
 
RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Hold

```java
PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
	.addType(PaymentType.HOLD)
	.addReasonCode(ReasonCode.OUT_OF_STOCK)
	.addOrderId("Order ID from original transaction")
	.addPaymentsReference("Pasref from original transaction");

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Release

```java
PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
	.addType(PaymentType.RELEASE)
	.addReasonCode(ReasonCode.FALSE_POSITIVE)
	.addOrderId("Order ID from original transaction")
	.addPaymentsReference("Pasref from original transaction");

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Receipt-In

```java
PaymentData paymentData = new PaymentData()
  	.addCvnNumber("123");

PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
 	.addType(PaymentType.RECEIPT_IN)
 	.addAmount(100)
 	.addCurrency("EUR")
 	.addPayerReference("payer ref for customer")
 	.addPaymentMethod("payment method ref for card")
 	.addPaymentData(paymentData);

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Payment-out

```java

PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
 	.addType(PaymentType.PAYMENT_OUT) 	
 	.addAmount(100)
 	.addCurrency("EUR")
 	.addPayerReference("payer ref for customer")
 	.addPaymentMethod("payment method ref for card")
 	.addRefundHash("SHA1 hash of refund password");

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Payer-new

```java

PayerAddress address = new PayerAddress()
    .addLine1("Apt 167 Block 10")
    .addLine2("The Hills")
    .addLine3("67-69 High St")
    .addCity("Hytown")
    .addCounty("Dunham")
    .addPostcode("3")
    .addCountryCode("IE")
    .addCountryName("Ireland");
    

Payer payer = new Payer()
    .addType("Business")
    .addRef("smithj01")
    .addTitle("Mr")
    .addFirstName("John")
    .addSurname("Smith")
    .addCompany("Acme")
    .addPayerAddress(address)
    .addHomePhoneNumber("+35317285355")
    .addWorkPhoneNumber("+35317433923")
    .addFaxPhoneNumber("+35317893248")
    .addMobilePhoneNumber("+353873748392")
    .addEmail("jsmith@acme.com")
    .addComment("Comment1")
    .addComment("Comment2");

PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
 	.addType(PaymentType.PAYER_NEW)  	
 	.addPayer(payer);

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Payer-edit

```java

PayerAddress address = new PayerAddress()
    .addLine1("Apt 167 Block 10")
    .addLine2("The Hills")
    .addLine3("67-69 High St")
    .addCity("Hytown")
    .addCounty("Dunham")
    .addPostcode("3")
    .addCountryCode("IE")
    .addCountryName("Ireland");
    

Payer payer = new Payer()
    .addType("Business")
    .addRef("smithj01")
    .addTitle("Mr")
    .addFirstName("John")
    .addSurname("Smith")
    .addCompany("Acme")
    .addPayerAddress(address)
    .addHomePhoneNumber("+35317285355")
    .addWorkPhoneNumber("+35317433923")
    .addFaxPhoneNumber("+35317893248")
    .addMobilePhoneNumber("+353873748392")
    .addEmail("jsmith@acme.com")
    .addComment("Comment1")
    .addComment("Comment2");

PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
 	.addType(PaymentType.PAYER_EDIT)  	
 	.addPayer(payer);

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Card-new

```java

Card card = new Card()
    .addReference("visa01")
    .addPayerReference("smithj01")
    .addNumber("420000000000000000")    
	.addExpiryDate("0119")	
	.addCardHolderName("Joe Smith")
	.addType(CardType.VISA)
	.addIssueNumber("1");

PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
 	.addType(PaymentType.CARD_NEW)  	
 	.addCard(card);

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Card-edit

```java

Card card = new Card()
    .addReference("visa01")
    .addPayerReference("smithj01")
    .addNumber("420000000000000000")    
	.addExpiryDate("0119")	
	.addCardHolderName("Joe Smith")
	.addType(CardType.VISA);

PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
 	.addType(PaymentType.CARD_UPDATE)  	
 	.addCard(card);

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```


### Card-delete

```java

Card card = new Card()
    .addReference("visa01")
    .addPayerReference("smithj01");

PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
 	.addType(PaymentType.CARD_CANCEL)  	
 	.addCard(card);

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Verify Card Enrolled

```java

PaymentData paymentData = new PaymentData()
  	.addCvnNumber("123");


ThreeDSecureRequest request = new ThreeDSecureRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
  	.addType(ThreeDSecureType.VERIFY_CARD_ENROLLED)
  	.addAmount(100)
  	.addCurrency("EUR")
  	.addPayerReference("payer ref from customer")
  	.addPaymentMethod("payment method ref from customer")
  	.addPaymentData(paymentData)
  	.addAutoSettle(new AutoSettle().addFlag(AutoSettleFlag.TRUE));
 	

RealexClient client = new RealexClient("shared secret");
ThreeDSecureResponse response = client.send(request);
```

### DCC Rate Lookup

```java

Card card = new Card()    
    .addNumber("420000000000000000")    
	.addExpiryDate("0119")	
	.addCardHolderName("Joe Smith")
	.addType(CardType.VISA);
	
DccInfo dccInfo = new DccInfo()
    .addDccProcessor("fexco");

PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
  	.addType(PaymentType.DCC_RATE_LOOKUP)
  	.addAmount(100)
  	.addCurrency("EUR")
  	.addCard(card)
  	.addDccInfo(dccInfo);

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```


### Authorisation with DCC Information

```java

Card card = new Card()    
    .addNumber("420000000000000000")    
	.addExpiryDate("0119")	
	.addCardHolderName("Joe Smith")
	.addType(CardType.VISA);
	
DccInfo dccInfo = new DccInfo()
    .addDccProcessor("fexco")
    .addRate(0.6868)
    .addAmount(13049)
    .addCurrency("GBP");

PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
  	.addType(PaymentType.DCC_AUTH)
  	.addAmount(19000)
  	.addCurrency("EUR")
  	.addCard(card)
  	.addDccInfo(dccInfo);

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Receipt-in OTB

```java
 
 PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
	.addType(PaymentType.RECEIPT_IN_OTB)
	.addPayerReference("payer ref from customer")
    .addPaymentMethod("payment method ref from customer");
	
RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);	
```

### DCC Stored Card Dcc Rate

```java
	
DccInfo dccInfo = new DccInfo()
    .addDccProcessor("fexco");

PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
  	.addType(PaymentType.STORED_CARD_DCC_RATE)
  	.addAmount(100)
  	.addCurrency("EUR")
 	.addPayerReference("payer ref for customer")
 	.addPaymentMethod("payment method ref for card")
  	.addDccInfo(dccInfo);

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Fraud Filter Request

```java

Card card = new Card()
  	.addExpiryDate("0119")
  	.addNumber("4242424242424242")
  	.addType(CardType.VISA)
  	.addCardHolderName("Joe Smith")
  	.addCvn("123");

PaymentRequest request = new PaymentRequest()
	.addMerchantId("Merchant ID")
	.addAccount("internet")
  	.addType(PaymentType.AUTH)
  	.addAmount(1000)
  	.addCurrency("EUR")
  	.addCard(card)
  	.addAutoSettle(new AutoSettle().addFlag(AutoSettleFlag.TRUE))
  	.addFraudFilter(new FraudFilter().addMode(FraudFilter.FraudFilterMode.PASSIVE));

RealexClient client = new RealexClient("shared secret");
PaymentResponse response = client.send(request);
```

### Fraud Filter Response

```java
// request is fraud filter
PaymentResponse response = client.send(request);

FraudFilter.FraudFilterMode mode = response.getFraudFilter().getMode();
FraudFilter.FraudFilterResult result = response.getFraudFilter().getResult();

List<FraudFilterRule> rules = response.getFraudFilter().getRules();

for (FraudFilterRule rule :rules ) {
    System.out.print(rule.getId());
    System.out.print(rule.getName());
    System.out.print(rule.getValue());
}
//or        
rules.get(0).getId();
```
## License
See the LICENSE file.


CHANGELOG.md
# Change Log
All notable changes to the SDK will be documented in this file.

## [1.3.2]
- Added new multi-capture flagging.

## [1.3.1]
- Updated hash for Mobile-Auth to support Google Pay.

## [1.3]
- Updated endpoints.

## [1.2]
- Added Card Storage: Receipt-in, Payment-out, Payer-New, Payer-Edit, Card-new, Card-update, Card-delete, Verify Enrolled Card, DCC Rate Lookup, DCC + Auth, Receipt in OTB

## [1.1]
- Added mobile payment type (auth-mobile) and relevant fields (mobile, token) to payment request.
- Added settle, void, rebate, OTB, credit, hold and release payment types.

## [1.0.1]
- Fixed CVN issue where leading 0s were being removed.

## [1.0]
- Initial release.


src/main/java/com/realexpayments/remote/sdk/domain/Payer.java
package com.realexpayments.remote.sdk.domain;

import com.realexpayments.remote.sdk.domain.payment.Comment;

import javax.xml.bind.annotation.*;
import java.util.ArrayList;
import java.util.List;

/**
 * <p>
 * Domain object representing Payer information to be passed to Realex.
 * </p>
 *
 * <p><code><pre>
 * Payer payer = new Payer()
 * .addType("Business")
 * .addRef("smithj01")
 * .addTitle("Mr")
 * .addFirstName("John")
 * .addSurname("Smith")
 * .addCompany("Acme")
 * .addAddress(address)
 * .addHomePhoneNumber("+35317285355")
 * .addWorkPhoneNumber("+35317433923")
 * .addFaxPhoneNumber("+35317893248")
 * .addMobilePhoneNumber("+353873748392")
 * .addEmail("jsmith@acme.com")
 * .addComment("Comment1")
 * .addComment("Comment2");
 *
 * </pre></code></p>
 *
 * @author vicpada
 *
 */
@XmlAccessorType(XmlAccessType.FIELD)
public class Payer {

    /**
     * The payer type can be used to identify the category of the Payer.
     * This can be defaulted to “Business”
     */
    @XmlAttribute(name = "type")
    private String type;

    /**
     * The payer ref is the reference for this customer. It must be unique.
     */
    @XmlAttribute(name = "ref")
    private String ref;

    /**
     * The payer’s title
     */
    @XmlElement(name = "title")
    private String title;

    /**
     * First name of payer.
     */
    @XmlElement(name = "firstname")
    private String firstName;

    /**
     * Surname of payer
     */
    @XmlElement(name = "surname")
    private String surname;

    /**
     * Company Name
     */
    @XmlElement(name = "company")
    private String company;

    /**
     * {@link PayerAddress} object containing the payer address.
     */
    @XmlElement(name = "address")
    private PayerAddress address;

    /**
     * {@link PhoneNumbers} object containing the payer phone numbers.
     */
    @XmlElement(name = "phonenumbers")
    private PhoneNumbers phoneNumbers;

    /**
     * The payer email
     */
    @XmlElement(name = "email")
    private String email;

    /**
     * List of {@link Comment} objects to be passed in request.
     * Optionally, up to two comments can be associated with any payer.
     */
    @XmlElementWrapper(name = "comments")
    @XmlElements(@XmlElement(name = "comment", type = Comment.class))
    private List<Comment> comments;

    /**
     * Constructor
     */
    public Payer() {
    }

    /**
     * Getter for type
     *
     * @return String
     */
    public String getType() {
        return type;
    }

    /**
     * Setter for type
     *
     * @param type
     */
    public void setType(String type) {
        this.type = type;
    }

    /**
     * Getter for ref
     *
     * @return String
     */
    public String getRef() {
        return ref;
    }

    /**
     * Setter for ref
     *
     * @param ref
     */
    public void setRef(String ref) {
        this.ref = ref;
    }

    /**
     *
     * Getter for title
     *
     * @return String
     */
    public String getTitle() {
        return title;
    }

    /**
     * Setter for title
     *
     * @param title
     */
    public void setTitle(String title) {
        this.title = title;
    }

    /**
     * Getter for firstName
     *
     * @return String
     */
    public String getFirstName() {
        return firstName;
    }

    /**
     * Setter for firstName
     *
     * @param firstName
     */
    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    /**
     * Getter for surname
     *
     * @return String
     */
    public String getSurname() {
        return surname;
    }

    /**
     * Setter for surname
     *
     *  @param surname
     */
    public void setSurname(String surname) {
        this.surname = surname;
    }

    /**
     * Getter for company
     *
     * @return String
     */
    public String getCompany() {
        return company;
    }

    /**
     * Setter for company
     *
     * @param company
     */
    public void setCompany(String company) {
        this.company = company;
    }

    /**
     * Getter for address
     *
     * @return Address
     */
    public PayerAddress getAddress() {
        return address;
    }

    /**
     * Setter for address
     *
     * @param address
     */
    public void setAddress(PayerAddress address) {
        this.address = address;
    }

    /**
     * Getter for phoneNumbers
     *
     * @return PhoneNumbers
     */
    public PhoneNumbers getPhoneNumbers() {
        return phoneNumbers;
    }

    /**
     * Setter for phoneNumbers
     *
     * @param phoneNumbers
     */
    public void setPhoneNumbers(PhoneNumbers phoneNumbers) {
        this.phoneNumbers = phoneNumbers;
    }

    /**
     * Getter for email
     *
     * @return String
     */
    public String getEmail() {
        return email;
    }

    /**
     * Setter for email
     *
     * @param email
     */
    public void setEmail(String email) {
        this.email = email;
    }

    /**
     * Getter for {@link Comment} list.
     *
     * @return List<Comment>
     */
    public List<Comment> getComments() {
        return comments;
    }

    /**
     * Setter for comments
     *
     * @param comments
     */
    public void setComments(List<Comment> comments) {
        this.comments = comments;
    }

    /**
     * Helper method for adding a payer home phone number.
     *
     * @param phoneNumber
     * @return Payer
     */
    public Payer addHomePhoneNumber(String phoneNumber) {

        //create new comments array list if null
        if (null == this.phoneNumbers) {
            this.phoneNumbers = new PhoneNumbers();
        }

        this.phoneNumbers.setHomePhoneNumber(phoneNumber);
        return this;
    }

    /**
     * Helper method for adding a payer work phone number.
     *
     * @param phoneNumber
     * @return Payer
     */
    public Payer addWorkPhoneNumber(String phoneNumber) {

        //create new comments array list if null
        if (null == this.phoneNumbers) {
            this.phoneNumbers = new PhoneNumbers();
        }

        this.phoneNumbers.setWorkPhoneNumber(phoneNumber);
        return this;
    }

    /**
     * Helper method for adding a payer fax phone number.
     *
     * @param phoneNumber
     * @return Payer
     */
    public Payer addFaxPhoneNumber(String phoneNumber) {

        //create new comments array list if null
        if (null == this.phoneNumbers) {
            this.phoneNumbers = new PhoneNumbers();
        }

        this.phoneNumbers.setFaxPhoneNumber(phoneNumber);
        return this;
    }

    /**
     * Helper method for adding a payer mobile phone number.
     *
     * @param phoneNumber
     * @return Payer
     */
    public Payer addMobilePhoneNumber(String phoneNumber) {

        //create new comments array list if null
        if (null == this.phoneNumbers) {
            this.phoneNumbers = new PhoneNumbers();
        }

        this.phoneNumbers.setMobilePhoneNumber(phoneNumber);
        return this;
    }


    /**
     * Helper method for adding a payer ref.
     *
     * @param ref
     * @return Payer
     */
    public Payer addRef(String ref) {
        this.ref = ref;
        return this;
    }

    /**
     * Helper method for adding a payer title.
     *
     * @param title
     * @return Payer
     */
    public Payer addTitle(String title) {
        this.title = title;
        return this;
    }

    /**
     * Helper method for adding a payer first name.
     *
     * @param firstName
     * @return Payer
     */
    public Payer addFirstName(String firstName) {
        this.firstName = firstName;
        return this;
    }

    /**
     * Helper method for adding a payer type.
     *
     * @param surname
     * @return Payer
     */
    public Payer addSurname(String surname) {
        this.surname = surname;
        return this;
    }

    /**
     * Helper method for adding a payer type.
     *
     * @param company
     * @return Payer
     */
    public Payer addCompany(String company) {
        this.company = company;
        return this;
    }

    /**
     * Helper method for adding a payer type.
     *
     * @param address
     * @return Payer
     */
    public Payer addPayerAddress(PayerAddress address) {
        this.address = address;
        return this;
    }

    /**
     * Helper method for adding a payer type.
     *
     * @param type
     * @return Payer
     */
    public Payer addType(String type) {
        this.type = type;
        return this;
    }

    /**
     * Helper method for adding a payer type.
     *
     * @param email
     * @return Payer
     */
    public Payer addEmail(String email) {
        this.email = email;
        return this;
    }

    /**
     * Helper method for adding a comment. NB Only 2 comments will be accepted by Realex.
     *
     * @param comment
     * @return Payer
     */
    public Payer addComment(String comment) {
        //create new comments array list if null
        if (null == this.comments) {
            this.comments = new ArrayList<Comment>();
        }

        int size = comments.size();
        this.comments.add(new Comment().addComment(comment).addId(++size));
        return this;
    }


}



